docs/bookstore/page-1/pg1-review-svc-1.md

```diff
@@ -0,0 +1,379 @@
+
+![Image](images/image12.png)
+
+Let's examine the `node-server/index.js` file, starting with the `/add` function. When a user submits a comment through the frontend, it is first received by this endpoint.
+
+```javascript
+app.post('/add', async (req, res) => {
+  try {
+    const receivedEvent = HTTP.toEvent({headers: req.headers, body: req.body});
+    const brokerURI = process.env.K_SINK;
+
+    if (receivedEvent.type === 'new-review-comment') {
+      const response = await fetch(brokerURI, {
+        method: 'POST',
+        headers: {
+          'Content-Type': 'application/cloudevents+json',
+          'ce-specversion': receivedEvent.specversion,
+          'ce-type': receivedEvent.type,
+          'ce-source': receivedEvent.source,
+          'ce-id': receivedEvent.id,
+        },
+        body: JSON.stringify(receivedEvent.data),
+      });
+    }
+  } catch (err) {
+    console.error(err);
+  }
+});
+```
+
+**Receiving Events**: The `/add` endpoint receives events from the frontend. It uses CloudEvents' SDK to convert the incoming request into a CloudEvent object:
+
+```javascript
+const receivedEvent = HTTP.toEvent({headers: req.headers, body: req.body});
+```
+
+**Determining Broker Address**: The broker's address is dynamically assigned in the cluster. The Node.js server retrieves this address from the environment variable `K_SINK`:
+
+```javascript
+const brokerURI = process.env.K_SINK;
+```
+
+You may wonder who told the environment variable about the address? That’s Knative SinkBinding.
+
+**Event Filtering**: The service checks the event type. If it's a `new-review-comment`, it forwards the event to the broker:
+
+```javascript
+if (receivedEvent.type === 'new-review-comment') {
+  // logic that forwards the event, see below
+}
+```
+
+**Forwarding Events Logic**: The event is forwarded to the broker with the appropriate CloudEvent headers:
+
+```javascript
+const response = await fetch(brokerURI, {
+  method: 'POST',
+  headers: {
+    'Content-Type': 'application/cloudevents+json',
+    'ce-specversion': receivedEvent.specversion,
+    'ce-type': receivedEvent.type,
+    'ce-source': receivedEvent.source,
+    'ce-id': receivedEvent.id,
+  },
+  body: JSON.stringify(receivedEvent.data),
+});
+```
+
+![Image](images/image21.png)
+
+**Exploring Other Functions**
+
+The Node.js server contains other functions that follow a similar pattern, with detailed comments explaining their functionality:
+
+- `/insert`: Receives CloudEvents and inserts the payload into the PostgreSQL database.
+- `/comment`: Creates a WebSocket connection with the frontend to transport comments from the database.
+
+### **Step 2: Create Broker**
+
+![Image](images/image17.png)
+
+The Broker acts as a router in your event-driven application, receiving events and routing them to the correct destination.
+
+
+
+- 1: Create a new YAML file named `node-server/config/200-broker.yaml` and add the following content:
```

Comment:
Why is the file named `200-broker` ? while this is used for our config files where there is a large number of YAML files, in this case, is not really needed as there is at most 3/4 YAMLs very different from each other